File: tools/apiClient.py
#!/usr/bin/env python3
"""apiClient.py - drive the IntegrityAnalysis REST service from Python.

Standard library only (no requests, no third-party packages), so it runs
anywhere Python 3.8+ does. The R twin is tools/apiClient.R; both were run
against a local service and the deployed one before the API User's Guide
(docs/api-users-guide.md) was written from their replies.

Usage:
    python tools/apiClient.py health  <base-url>
    python tools/apiClient.py parse   <base-url> <file>
    python tools/apiClient.py analyze <base-url> <file> [--seed N]

<file> is an article PDF, a Word manuscript (.docx), a JATS XML article
(.xml), a spreadsheet (csv/xls/xlsx), or a picture of a table
(jpg/png/tif); one file per call. The service takes no replication count:
every row runs the app's staged Monte Carlo (1,000 / 10,000 / 100,000
replicates, escalating only while the row alarms). The bearer token comes
from the
INTEGRITY_API_TOKEN environment variable; /health needs none. Written
2026-09-03 by Claude Code (model Claude Fable 5.1) at Steve Shafer's
request, beside the R client.
"""
import csv
import io
import json
import mimetypes
import os
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid


def usage():
    print(__doc__.split("Usage:")[1].split("<file>")[0].strip())
    sys.exit(2)


# THE CLIENT DOES NOT FOLLOW REDIRECTS, AND A REMOTE SERVICE MUST BE HTTPS
# (outside security review, 2026-09-26). Python's default opener follows a
# 3xx wherever it points and re-sends the request's headers there, so a
# redirecting endpoint - or anything between the client and it - would
# receive the bearer token, and a plain http:// base URL would carry the
# token and the manuscript in clear. A redirect is reported as a wrong
# base URL and nothing is re-sent; http:// is accepted only for a service
# on this machine (the guide's local-run examples).
class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None                         # urlopen raises HTTPError(3xx)


_opener = urllib.request.build_opener(_NoRedirect)
_LOCAL_HOSTS = ("127.0.0.1", "localhost", "::1", "[::1]")


def check_base(base):
    u = urllib.parse.urlsplit(base)
    host = (u.hostname or "").lower()
    if u.scheme == "https" or (u.scheme == "http" and host in _LOCAL_HOSTS):
        return
    print("the service URL must be https:// (http:// only for a service on this machine): %s" % base)
    sys.exit(2)


def redirected(e):
    """A 3xx from the service: say where it pointed, send nothing there."""
    if 300 <= e.code < 400:
        print("the service answered %d with a redirect to %s; this client does not follow "
              "redirects with your token - check the base URL with the operator"
              % (e.code, e.headers.get("Location", "(no Location header)")))
        return True
    return False


def health(base):
    try:
        with _opener.open(base + "/health", timeout=30) as r:
            body = json.loads(r.read().decode("utf-8"))
            build = body.get("commit")
            print("health: %d  ok=%s%s%s" % (
                r.status, body.get("ok"),
                "  build %s" % build[:8] if build else "",
                "  %s" % body["engine"] if body.get("engine") else ""))
            return True
    except urllib.error.HTTPError as e:
        if redirected(e):
            return False
        print("health: %s answered HTTP %d" % (base, e.code))
        return False
    except Exception as e:  # noqa: BLE001 - report and stop
        print("health: could not reach %s - %s" % (base, e))
        return False


def multipart(fields, file_field, path):
    """Build a multipart/form-data body by hand: fields + one file."""
    boundary = "----IntegrityAnalysis" + uuid.uuid4().hex
    ctype = mimetypes.guess_type(path)[0] or "application/octet-stream"
    parts = []
    for k, v in fields.items():
        parts.append(("--%s\r\nContent-Disposition: form-data; name=\"%s\"\r\n\r\n%s\r\n"
                      % (boundary, k, v)).encode("utf-8"))
    with open(path, "rb") as fh:
        data = fh.read()
    parts.append(("--%s\r\nContent-Disposition: form-data; name=\"%s\"; filename=\"%s\"\r\n"
                  "Content-Type: %s\r\n\r\n" % (boundary, file_field, os.path.basename(path), ctype)).encode("utf-8"))
    parts.append(data)
    parts.append(("\r\n--%s--\r\n" % boundary).encode("utf-8"))
    return b"".join(parts), "multipart/form-data; boundary=" + boundary


def formula_cells(text):
    """Count the cells a spreadsheet would read as a formula: a cell that
    begins with =, @, a tab or a carriage return, or with + or - and then
    something other than a number ("-0.5" is a number; "-cmd" is not)."""
    n = 0
    for row in csv.reader(io.StringIO(text)):
        for cell in row:
            if not cell:
                continue
            c = cell[0]
            if c in "=@\t\r":
                n += 1
            elif c in "+-" and not cell[1:2].isdigit() and cell[1:2] != ".":
                n += 1
    return n


def save_csv(stem, suffix, text):
    if not text:
        return
    out = "%s-%s.csv" % (stem, suffix)
    with open(out, "w", encoding="utf-8", newline="") as fh:
        fh.write(text)
    print("  wrote %s (%d rows)" % (out, text.count("\n") - 1))
    # THE TEMPLATE IS SAVED VERBATIM, AND SAID SO WHEN THAT MATTERS (outside
    # security review, 2026-09-26). templateCsv is the round-trip payload -
    # the service must accept it back unchanged, so the client cannot
    # sanitise its labels as the results CSV's are - but it is written as
    # an ordinary .csv beside the input, and a label a spreadsheet would
    # read as a formula ("=...", "@...", "+text") is a manuscript's own
    # text. When such a cell is present the user is told, and told to edit
    # the file in a text editor rather than open it in spreadsheet software.
    if suffix == "template":
        k = formula_cells(text)
        if k:
            print("  note: %d cell(s) in %s begin with a character spreadsheet software reads as a "
                  "formula (=, +, -, @); the file is kept exactly as the service returned it so it "
                  "can be sent back - edit it in a text editor, not in a spreadsheet" % (k, out))


def main(argv):
    seed = None
    if "--seed" in argv:                      # a Monte Carlo seed: same file, seed and build -> same numbers
        k = argv.index("--seed")
        if k + 1 >= len(argv):
            usage()
        seed = argv[k + 1]
        argv = argv[:k] + argv[k + 2:]
    if len(argv) < 2 or argv[0] not in ("health", "parse", "analyze"):
        usage()
    verb, base = argv[0], argv[1].rstrip("/")
    check_base(base)
    if not health(base) or verb == "health":
        sys.exit(0 if verb == "health" else 1)
    if len(argv) < 3:
        usage()
    token = os.environ.get("INTEGRITY_API_TOKEN", "")
    if not token:
        print("set INTEGRITY_API_TOKEN to the token the operator issued you")
        sys.exit(2)
    path = argv[2]
    if not os.path.exists(path):
        print("no such file: %s" % path)
        sys.exit(2)
    fields = {}
    url = base + "/" + verb
    if seed is not None and verb == "analyze":
        # on the URL, not as a form part: a text part without a
        # Content-Type is dropped by the service's multipart parser
        # (found 2026-09-05), and the query string always arrives
        url += "?seed=" + urllib.parse.quote(str(seed), safe="")
    body, ctype = multipart(fields, "file", path)
    req = urllib.request.Request(url, data=body, method="POST",
                                 headers={"Authorization": "Bearer " + token,
                                          "Content-Type": ctype,
                                          "Content-Length": str(len(body))})
    t0 = time.time()
    try:
        with _opener.open(req, timeout=900 if verb == "analyze" else 300) as r:
            status, raw = r.status, r.read()
    except urllib.error.HTTPError as e:      # 4xx/5xx still carry a JSON body
        if redirected(e):
            sys.exit(1)
        status, raw = e.code, e.read()
    except Exception as e:  # noqa: BLE001
        print("%s: request failed - %s" % (verb, e))
        sys.exit(1)
    secs = round(time.time() - t0, 1)
    print("%s %s: HTTP %d in %s s" % (verb, os.path.basename(path), status, secs))
    if os.environ.get("INTEGRITY_API_SAVE_RAW"):      # the reply exactly as received
        rawpath = "%s-%s-reply.json" % (os.path.splitext(path)[0], verb)
        with open(rawpath, "wb") as fh:
            fh.write(raw)
        print("  raw reply: %s" % rawpath)
    try:
        b = json.loads(raw.decode("utf-8"))
    except ValueError:
        print("  (no JSON body) %s" % raw[:300])
        sys.exit(1)
    print("  ok=%s  deleted=%s%s" % (b.get("ok"), b.get("deleted"),
                                    "  engine=%s" % b["engine"] if b.get("engine") else ""))
    for key in ("reasons", "flags"):
        if b.get(key):
            v = b[key] if isinstance(b[key], list) else [b[key]]   # unboxed JSON: one reason is a string
            print("  %s: %s" % (key, "; ".join(str(x) for x in v)))
    if b.get("rows") is not None:
        print("  rows: %s" % b["rows"])
    for s in b.get("skipped") or []:
        print("    - %s: %s" % (s.get("label"), s.get("reason")))
    if b.get("trials") is not None:
        print("  trials: %s" % b["trials"])
    if b.get("overallP") is not None:
        print("  overall p: %s" % b["overallP"])
    stem = os.path.splitext(path)[0]
    save_csv(stem, "template", b.get("templateCsv"))
    save_csv(stem, "results", b.get("resultsCsv"))
    for name, csv in (b.get("journalTables") or {}).items():
        save_csv(stem, "journal-" + "".join(c if c.isalnum() or c in "._-" else "_" for c in name), csv)
    if b.get("journalTablesOmitted"):
        print("  journal tables omitted: %s" % b["journalTablesOmitted"])
    sys.exit(0 if b.get("ok") else 1)


if __name__ == "__main__":
    main(sys.argv[1:])

File: tools/apiClient.R
# apiClient.R - drive the IntegrityAnalysis REST service from R, by hand.
#
############################################################################
# Provenance                                                               #
# Written 2026-09-03 by Claude Code (model Claude Fable 5.1) at Steve      #
# Shafer's request: "I haven't tested the API interface ... Have you       #
# written an R script that you've used to test the API interface?" The    #
# test suite (tests/testthat/test-api-service.R) boots the service in a   #
# subprocess and posts files over HTTP, but it is a test, not something   #
# a person runs. This is the by-hand client: health, then one file to     #
# /parse or /analyze, the reply printed and its CSVs saved beside the     #
# input. It uses only packages the package already imports (httr2, curl).#
############################################################################
#
# Usage (from anywhere; Rscript from the R 4.5.3 installation):
#
#   Rscript tools/apiClient.R health  <base-url>
#   Rscript tools/apiClient.R parse   <base-url> <file>
#   Rscript tools/apiClient.R analyze <base-url> <file>
#
#   <file>     an article PDF, Word manuscript (.docx), JATS XML (.xml),
#              spreadsheet (csv/xls/xlsx), or picture of a table
#              (jpg/png/tif). One file per call - the service takes no zip.
#   The service takes no replication count: every row runs the app's
#   staged Monte Carlo (1,000 / 10,000 / 100,000 replicates, escalating
#   only while the row alarms), and the M column of the results says
#   what each row used.
#
# The bearer token comes from the INTEGRITY_API_TOKEN environment variable
# (the plaintext token the operator issued with tools/issueApiToken.R;
# the service holds only its hash). /health needs no token.
#
# AGAINST A LOCAL SERVICE, in a second R session (or terminal):
#   Sys.setenv(INTEGRITY_API_TOKENS = "local-test-token")
#   IntegrityAnalysis::runApiService(port = 8080, host = "127.0.0.1")
# then here:
#   Sys.setenv(INTEGRITY_API_TOKEN = "local-test-token") / set INTEGRITY_API_TOKEN=...
#   Rscript tools/apiClient.R analyze http://127.0.0.1:8080 path/to/article.pdf
#
# AGAINST THE DEPLOYED SERVICE (AWS App Runner, service
# "integrityanalysis-api"): the base URL is the service's ServiceUrl -
#   aws apprunner list-services --profile steve --region us-east-1 \
#     --query "ServiceSummaryList[?ServiceName=='integrityanalysis-api'].ServiceUrl" --output text
# (prefix https://), and the token is one issued to you. Everything you
# send is processed in a per-request directory and deleted when the
# reply is written; the reply says so ("deleted": true).

args <- commandArgs(trailingOnly = TRUE)
usage <- function() {
  cat("usage: Rscript tools/apiClient.R health  <base-url>\n",
      "       Rscript tools/apiClient.R parse   <base-url> <file>\n",
      "       Rscript tools/apiClient.R analyze <base-url> <file>\n", sep = "")
  quit(status = 2)
}
if (length(args) < 2) usage()
verb <- args[1]; base <- sub("/+$", "", args[2])
if (!verb %in% c("health", "parse", "analyze")) usage()
if (verb != "health" && length(args) < 3) usage()
suppressPackageStartupMessages({ library(httr2); library(curl) })
token <- Sys.getenv("INTEGRITY_API_TOKEN", "")

say <- function(...) cat(paste0(..., collapse = ""), "\n")
show <- function(label, x) if (!is.null(x) && length(x)) say("  ", label, ": ", paste(unlist(x), collapse = "; "))

# ---- health: open, no token ------------------------------------------------
h <- tryCatch(request(paste0(base, "/health")) |> req_timeout(30) |> req_perform(),
              error = function(e) e)
if (inherits(h, "error")) {
  say("health: could not reach ", base, " - ", conditionMessage(h)); quit(status = 1)
}
hb <- resp_body_json(h)
say("health: ", resp_status(h), "  ok=", isTRUE(hb$ok),
    if (!is.null(hb$commit)) paste0("  build ", substr(hb$commit, 1, 8)) else "",
    if (!is.null(hb$engine)) paste0("  ", hb$engine) else "")
if (verb == "health") quit(status = 0)

# ---- parse / analyze: one file, bearer token ------------------------------
if (!nzchar(token)) { say("set INTEGRITY_API_TOKEN to the token the operator issued you"); quit(status = 2) }
file <- args[3]
if (!file.exists(file)) { say("no such file: ", file); quit(status = 2) }

req <- request(paste0(base, "/", verb)) |>
  req_headers(Authorization = paste("Bearer", token)) |>
  req_timeout(if (verb == "analyze") 900 else 300) |>
  req_error(is_error = function(resp) FALSE)
body <- list(file = form_file(file))
req <- req_body_multipart(req, !!!body)

t0 <- Sys.time()
r <- tryCatch(req_perform(req), error = function(e) e)
secs <- round(as.numeric(difftime(Sys.time(), t0, units = "secs")), 1)
if (inherits(r, "error")) { say(verb, ": request failed - ", conditionMessage(r)); quit(status = 1) }
status <- resp_status(r)
# INTEGRITY_API_SAVE_RAW=1 keeps the reply exactly as received, beside the
# input - for a support question, or for reading the fields the guide names
if (nzchar(Sys.getenv("INTEGRITY_API_SAVE_RAW", ""))) {
  raw <- paste0(tools::file_path_sans_ext(file), "-", verb, "-reply.json")
  writeLines(resp_body_string(r), raw); say("  raw reply: ", raw)
}
b <- tryCatch(resp_body_json(r), error = function(e) NULL)
say(verb, " ", basename(file), ": HTTP ", status, " in ", secs, " s")
if (is.null(b)) { say("  (no JSON body) ", resp_body_string(r)); quit(status = 1) }

say("  ok=", isTRUE(b$ok), "  deleted=", isTRUE(b$deleted),
    if (!is.null(b$engine)) paste0("  engine=", b$engine) else "")
show("reasons", b$reasons)
show("flags", b$flags)
if (!is.null(b$rows)) say("  rows: ", b$rows)
if (length(b$skipped)) {
  say("  skipped ", length(b$skipped), " line(s):")
  for (s in b$skipped) say("    - ", s$label, ": ", s$reason)
}
if (!is.null(b$trials)) say("  trials: ", b$trials)
if (!is.null(b$overallP)) say("  overall p: ", b$overallP)

# the CSVs the reply carries, saved beside the input
stem <- tools::file_path_sans_ext(file)
saveCsv <- function(txt, suffix) {
  if (is.null(txt) || !nzchar(txt)) return(invisible())
  out <- paste0(stem, "-", suffix, ".csv")
  writeLines(txt, out)
  say("  wrote ", out, " (", length(strsplit(txt, "\n")[[1]]) - 1L, " rows)")
}
saveCsv(b$templateCsv, "template")
saveCsv(b$resultsCsv, "results")
# journalTables: one CSV per trial, keyed by trial name (the reconstructed
# journal-style table, for comparison against the manuscript page)
if (length(b$journalTables)) {
  nms <- names(b$journalTables); if (is.null(nms)) nms <- paste0("trial", seq_along(b$journalTables))
  for (i in seq_along(b$journalTables))
    saveCsv(b$journalTables[[i]], paste0("journal-", gsub("[^A-Za-z0-9._-]", "_", nms[i])))
}
if (length(b$journalTablesOmitted) && nzchar(b$journalTablesOmitted[[1]]))
  say("  journal tables omitted: ", b$journalTablesOmitted[[1]])
quit(status = if (isTRUE(b$ok)) 0 else 1)
